File: lab9_regressao_logistica.R
# Title     : Regressão Logística -> Sucesso ou Fracasso
# Objective : Conhecer as ferramentas de R que podem nos apoiar no processo de Regressão Logística
# Created by: accol
# Created on: 29/05/2020

# Para este laboratório usaremos base de dados de campanha eleitoral -> eleicao.csv e novoscandidatos.csv
# Nesta prática queremos a partir dos dados da base eleicao.csv queremos prever se os candidados de novoscandidatos.csv serão ou não eleitos
# Na tabela eleicao.csv teremos o nome do candidato, a situação -> se ele foi eleito ou não e despesas -> quanto ele investiu na campanha


# Importando dados no R -> observe o uso da opção file.chose() => ela abrairá uma caixa de diálogo na qual você poderá selecionar o arquivo desejado. Note também que foi necessário definir o separador do arquivo e se há ou não um header >- acompanhe. Essa opção foi deixada como comentário para que você saiba que pode contar com ela em caso de necessidade, na prática usaremos a leitura direta do diretório

#eleicao = read.csv(file.choose(), sep = ';', header = T)
eleicao = read.csv('D:/Users/Angelo/AULAS/Dados/Eleicao.csv', sep = ';', header = T)
#fix(eleicao) # -> é uma forma de exibir a tabela caso esteja numa ide sem muitos recursos
# Vamos fazer uma inspeção gráfica na base de dados -> analisando a situação (eleito ou não) com as despesas (investimento em campanha)
plot(eleicao$DESPESAS, eleicao$SITUACAO)
# Vamos agora gerar um sumário dos nossos dados
summary(eleicao)
# Vamos agora verificar se existe uma correlação entre a Situação e as Despesas
analise1 = cor(eleicao$DESPESAS, eleicao$SITUACAO)
cat('\nA correlação entre as despesas (investimento) e a situação (eleito ou não) é:\n', analise1)
# Observe que a correlação é de 81% -> portanto, há uma boa correlação entre as variáveis Despesas e Situação
# Assim podemos agora passar para a criação do nosso modelo vamos usar a função glm() que diferente do lm() cria vários tipos de regressões lineares -> sendo que o Tipo devemos passar por parâmetro -> no caso vamos usar family='binomial' para trabalharmos com regressão logística
modelo = glm(SITUACAO ~ DESPESAS, data = eleicao, family = 'binomial')
# criado o modelo podemos visualizar seu sumário
summary(modelo)
print(modelo)
# Vamos fazer uma inspeção gráfica na base de dados -> analisando a situação (eleito ou não) com as despesas (investimento em campanha) -> formatando um pouco o gráfico anterior dando melhor visualização
plot(eleicao$DESPESAS, eleicao$SITUACAO, col = 'red', pch = 20)
# Agora neste mesmo gráfico vamos gerar alguns pontos que são ajustes feito pelo modelo (modelo$fitted)
points(eleicao$DESPESAS, modelo$fitted.values, col = 'yellow', pch = 4)
# Neste gráfico poderá observar o efeito da binomial utilizada para realizar a regressão logística
# Ajustado o modelo vamos fazer uso dele para prever a SITUAÇÃO dos novos candidatos -> vamos ler nossa segunda base de dados novoscandidatos.csv
prever_eleicao = read.csv('D:/Users/Angelo/AULAS/Dados/novoscandidatos.csv', sep = ';', header = T)
#fix(prever_eleicao) # -> é uma forma de exibir a tabela caso esteja numa ide sem muitos recursos
# Vamos agora gerar um sumário dos nossos dados
summary(prever_eleicao)
# Vamos agora fazer uma previsão (SITUAÇÃO -> ELEITO OU NÃO) a partir do quanto os candidatos desejam investir na campanha
# Vamos criar uma nova coluna RESULT e chamar a função predict >- acompanhe
prever_eleicao$RESULT = predict(modelo, newdata = prever_eleicao, type = 'response')
# fix(prever_eleicao) # Visualiza de forma tabular os dados
cat('\nOlhando para o que foi previsto \n', prever_eleicao$RESULT)
# Podemos observar que os dois últimos candidatos possuem uma grande chance de serem eleitos -> acima de 94%

File: lab9_regressao_logistica.py
# Title     : Regressão Logística -> Sucesso ou Fracasso
# Objective : Conhecer as ferramentas de Python que podem nos apoiar no processo de Regressão Logística
# Created by: accol
# Created on: 29/05/2020

# Para este laboratório usaremos base de dados de campanha eleitoral -> eleicao.csv e novoscandidatos.csv
# Nesta prática queremos a partir dos dados da base eleicao.csv queremos prever se os candidados de novoscandidatos.csv serão ou não eleitos
# Na tabela eleicao.csv teremos o nome do candidato, a situação -> se ele foi eleito ou não e despesas -> quanto ele investiu na campanha


import pandas as pd 
import matplotlib.pyplot as plt
import numpy as np 
from sklearn.linear_model import LogisticRegression


# Lendo nossa base de dados e fazendo uma primeira inspeção básica. Atenção este arquivo .csv usa ';' como separador => será necessário especificar
base = pd.read_csv('D:/Users/Angelo/AULAS/Dados/Eleicao.csv', sep=';')
print('\nA Dimensão de nossa base Eleicao.csv é \n', base.shape)
print('\nUm resumo dos dados de base Eleicao.csv é \n', base.describe())
print('\nOlhando como os dados são organizados na base Eleicao.csv\n', base.head())
# Vamos fazer uma inspeção visual em nosso DataFrame. Note que, neste caso estaremos utilizando uma formadiferente de acesso a um determinado atributo em um DataFrame
plt.scatter(base.DESPESAS, base.SITUACAO)
plt.show()
# Agora vamos calcular a Correlação entre DESPESAS e SITUACAO
correlacao = np.corrcoef(base.DESPESAS, base.SITUACAO)
print('\nA correlação entre os atributos DESPESAS e SITUACAO são \n', correlacao)
# Observe que enocntramos uma correlaçao positiva forte de 81.2% -> indicativo de uma excelente correlação => viabilizando o processo de regressão linear logistica -> continuemos
# Nosso próximo passo será a criação do nosso modelo de Regressão Logistica . Não se esqueça de criar as variáveis X (INVESTIMENTO) e y (DESPESAS) -> em seguida vamos criar e treinar o modelo com o método .fit(). Lembre de fazer o .reshape() na variável X
X = base.iloc[:,2].values
# Note que nosso rechape será um pouco diferente -> vale o aprendizado => observe como podemos criar um novo eixo na nossa variável X
X = X[:, np.newaxis]
y = base.iloc[:, 1].values
# Chegou o momento de efetivamente criar o modelo
modelo = LogisticRegression()
modelo.fit(X, y)
# Vamos fazer uma inspeção em nosso modelo -> queremos conhecer seus parâmetros chaves => coef_ e intercept_
print('\nO coeficiente de modelo é:\n', modelo.coef_)
print('\nO ponto de interceptação de modelo é:\n', modelo.intercept_)
# Fazendo uma inspeção visual
plt.scatter(X, y)
# Vamos agora preparar o ambiente para uma simulação -> usaremos 100 registros. Para essa simulação criaremos uma variável X_teste que receberá 100 registros iniciando em 10 e finalizando em 3000 (intervalo da coluna DESPESAS). Faremos uso do método linspace() acompanhe
X_teste = np.linspace(10, 3000, 100)
print('\nQuem é nossa variável X_teste? \n', X_teste)
# Usaremos X-teste para fazermos uma previsão com nosso modelo. Acompanhe -> criaremos uma função para essa ação +> observe o uso de uma sgmoide muito comum em redes Neurais
def model(x):
    return 1 / (1 + np.exp(-x))
# Vamos criar agora uma variável r de resultado para a qual passamos os valores -> observe o uso do método .ravel() que vai transformar o numpy array do formato de matriz para o formato de vetor. r nos fornece o erro do modelo
r = model(X_teste * modelo.coef_ + modelo.intercept_).ravel()
# Inspecionado visulamente
plt.plot(X_teste, r, color = 'red')
plt.show()
# Modelo preparado e testado vamos agora usar o modelo para prever o que acontecerá com os candidatos presentes na base NovosCandidatos.csv -> queremos prever se serão ou não eleitos
# Precisamos carregar a nova base de dados
base_previsoes = pd.read_csv('D:/Users/Angelo/AULAS/Dados/NovosCandidatos.csv', sep=';')
# Vamos criar uma variável despesas -> não se esqueça do reshape()
despesas = base_previsoes.iloc[:, 1].values
despesas = despesas.reshape(-1, 1)
# Podemos agora realizar nossa previsão
previsoes_teste = modelo.predict(despesas)
print('\nOs candidatos com grande chance de serem eleitos são:\n', previsoes_teste)
# Para melhor visualizar podemos usar o seguinte:
base_previsoes = np.column_stack((base_previsoes, previsoes_teste))
print('\nVisualizando os candidatos e sua classificaçao em razão de suas despesas de campanha\n', base_previsoes)
